src/pg_copy.c

```diff
@@ -0,0 +1,574 @@
+{
+	return &cs->line_buf;
+}
+
+#define MAX_SHARDS 1001
+
+static uint32 shard_id_hash_fn(const void *key, Size keysize)
+{
+	return *(int32*)key ^ *((int32*)key+1);
+}
+
+/* 
+ * Construct hash table used for shardId->Connection mapping.
+ * We can not use connection cache from connection.c used by GteConnection because
+ * we need to establish multiple connections with each nodes: one connection per shard
+ */
+static HTAB *
+CreateShardToConnectionHash()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ShardId);
+	info.entrysize = sizeof(ShardConnections);
+	info.hash = shard_id_hash_fn;
+
+	return hash_create("shardToConn", MAX_SHARDS, &info, HASH_ELEM | HASH_FUNCTION);
+}
+
+static char const* ConstructCopyStatement(CopyStmt *copyStatement, ShardId shardId)
+{
+    StringInfo buf = makeStringInfo();
+    ListCell* cell;
+    char sep;
+    char const* qualifiedName = quote_qualified_identifier(copyStatement->relation->schemaname,
+                                                           copyStatement->relation->relname);
+    appendStringInfo(buf, "COPY %s_%ld ", qualifiedName, (long)shardId);    
+    if (copyStatement->attlist) { 
+        sep = '(';
+        foreach(cell, copyStatement->attlist)
+        {
+            appendStringInfoChar(buf, sep);
+            appendStringInfoString(buf, strVal(lfirst(cell)));
+            sep = ',';
+        }
+        appendStringInfoChar(buf, ')');        
+    }
+    appendStringInfoString(buf, "FROM STDIN");
+
+    if (copyStatement->options) { 
+        appendStringInfoString(buf, " WITH ");
+        sep = '(';
+        foreach(cell, copyStatement->options)
+        {
+            DefElem* def = (DefElem *) lfirst(cell);     
+            appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+            sep = ',';
+        }
+        appendStringInfoChar(buf, ')');        
+    }   
+    return buf->data;
+}
+
+static bool PgCopyPrepareTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	PQputCopyEnd(conn, NULL);
+	return tmgr->Prepare(conn, relationName, shardId);
+}
+
+static bool PgCopyAbortTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	if (status) 
+	{
+		tmgr->RollbackPrepared(conn, relationName, shardId);
+	} else {
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+
+static bool PgCopyEndTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	
+	if (status)
```

Comment:
I'd expect we only get here if prepare was sent on all connections so can never be false. Maybe add an Assert.